pkg/utils/taints.go

```diff
@@ -0,0 +1,120 @@
+package utils
+
+// Copy paste from SNR - https://github.com/medik8s/self-node-remediation/blob/main/pkg/utils/taints.go
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	farV1alpha1 "github.com/medik8s/fence-agents-remediation/api/v1alpha1"
+)
+
+var (
+	loggerTaint           = ctrl.Log.WithName("taints")
+	FARNodeNoExecuteTaint = &corev1.Taint{
+		Key:    farV1alpha1.Medik8sNoExecuteTaintKey,
+		Value:  farV1alpha1.FARNoExecuteTaintKeyValue,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+)
+
+// taintExists checks if the given taint exists in list of taints. Returns true if exists false otherwise.
+func taintExists(taints []corev1.Taint, taintToFind *corev1.Taint) bool {
+	for _, taint := range taints {
+		if taint.MatchTaint(taintToFind) {
+			return true
+		}
+	}
+	return false
+}
+
+// deleteTaint removes all the taints that have the same key and effect to given taintToDelete.
+func deleteTaint(taints []corev1.Taint, taintToDelete *corev1.Taint) ([]corev1.Taint, bool) {
+	var newTaints []corev1.Taint
+	deleted := false
+	for i := range taints {
+		if taintToDelete.MatchTaint(&taints[i]) {
+			deleted = true
+			continue
+		}
+		newTaints = append(newTaints, taints[i])
+	}
+	return newTaints, deleted
+}
+
+// HandleTaint calls other taint-functions basesd on the taint name and acion.
+// If there is no associated node for FAR or taint name is invalid or taint action is invalid then it returns an error
+func HandleTaint(r client.Client, taintName, actionOnTaint string, far *farV1alpha1.FenceAgentsRemediation) error {
+	var taintObj *corev1.Taint
+	// check if taint exist
+	if taintName == farV1alpha1.FARNoExecuteTaint {
+		taintObj = FARNodeNoExecuteTaint
+	} else {
+		return fmt.Errorf("invalid taintName - %s", taintName)
+	}
+
+	// find node
+	node, err := getNodeFromFAR(r, far)
+	if err != nil {
+		return err
+	}
+
+	if actionOnTaint == farV1alpha1.AppendTaintAction {
+		return appendTaint(r, taintObj, node)
+	} else if actionOnTaint == farV1alpha1.RemoveTaintAction {
+		return removeTaint(r, taintObj, node)
+	} else {
+		return fmt.Errorf("invalid action for taint - %s", actionOnTaint)
+	}
+}
+
+// appendTaint appends new taint to the taint list when it is not present, and returns error if it fails in the process
+func appendTaint(r client.Client, taintObj *corev1.Taint, node *corev1.Node) error {
+
+	if taintExists(node.Spec.Taints, taintObj) {
+		return nil
+	}
+
+	// add the taint to the taint list
+	patch := client.MergeFrom(node.DeepCopy())
+	taint := *taintObj
```

Comment:
sorry if it is obvious, but why this assignment?

Reply:
To store and declare the taint object in a different place (in the beginning), and in the function we just do assignment.

Reply:
This is indeed confusing.

IIUC `getTaintAndNode` is returning  a singleton taint which shouldn't be modified, therefore enforcing this assignment.

I think this is wrong, I suggest instead that `getTaintAndNode` would return a copy of the taint which in turn could be modified.

Reply:
- getTaintAndNode should not do two unrelated things in one function in first place IMHO
- getting a taint by name, which exist as pointer in a global var, is irritating IMHO... why not just a createNoExecuteTaint function, which builds and returns the taint by value and not pointer?